contracts/Committee.sol

```diff
@@ -166,374 +233,479 @@ contract Committee is ICommittee, Ownable {
-				_evict(lowestWeightPos); // evict lowest weight
-			}
-			_onTopologyModification();
-			newStandbySize = maxStandbys;
-		}
+					isStandby: false,
+					inCommittee: true
+				})
+			}),
+			_settings
+		) >= 0;
 	}
 
-	function _adjustPositionInTopology(uint pos) private returns (bool committeeChanged, bool standbysChanged) {
-		// TODO if a validator leaves committee it may be replaced by a timed-out, ready-for-committee standby
-		(uint newPos, uint prevCommitteeSize, uint newCommitteeSize, uint newStandbySize) = _repositionTopologyMember(pos);
+	function outranksLowestCommitteeMember(Member memory member, CommitteeInfo memory _committeeInfo, Settings memory _settings) private view returns (bool) {
+		return _committeeInfo.committeeSize > 0 && compareMembersPerCommitteeCriteria(
+			member,
+			Member({
+				addr: _settings.minimumAddress,
+				data: MemberData({
+					isMember: true,
+					readyForCommittee: true,
+					readyToSyncTimestamp: now,
+					weight: _committeeInfo.minCommitteeMemberWeight,
 
-		bool inCommitteeBefore = pos < prevCommitteeSize;
-		bool isStandbyBefore = !inCommitteeBefore;
+					isStandby: false,
+					inCommittee: true
+				})
+			}),
+			_settings
+		) > 0;
+	}
 
-		bool inCommitteeAfter = newPos < newCommitteeSize;
-		bool isStandbyAfter = !inCommitteeAfter;
+	function outranksLowestStandby(Member memory member, CommitteeInfo memory _committeeInfo, Settings memory _settings) private view returns (bool) {
+		return _committeeInfo.standbysCount > 0 && compareMembersPerStandbyCriteria(
+			member,
+			Member({
+				addr: _committeeInfo.minStandbyAddress,
+				data: MemberData({
+					isMember: true,
+					readyForCommittee: false,
+					readyToSyncTimestamp: now,
+					weight: _committeeInfo.minStandbyWeight,
+
+					isStandby: true,
+					inCommittee: false
+				})
+			}),
+		_settings) > 0;
+	}
+
+	function qualifiesForCommittee(Member memory member, Settings memory _settings, uint committeeSize, bool _outranksLowestCommitteeMember) private view returns (bool) {
+		return (
+			member.data.isMember &&
+			member.data.weight > 0 &&
+			member.data.readyForCommittee &&
+			!isReadyToSyncStale(member.data.readyToSyncTimestamp, member.data.inCommittee, _settings) &&
+			(
+				_outranksLowestCommitteeMember ||
+				_settings.minCommitteeSize > 0 && committeeSize < _settings.minCommitteeSize ||
+				committeeSize < _settings.maxCommitteeSize && isAboveCommitteeEntryThreshold(member, _settings)
+			)
+		);
+	}
+
+	function qualifiedToJoinCommittee(Member memory member, Settings memory _settings, CommitteeInfo memory _committeeInfo) private view returns (bool) {
+		return qualifiesForCommittee(member, _settings, _committeeInfo.committeeSize, outranksLowestCommitteeMember(member, _committeeInfo, _settings));
+	}
+
+	function joinMemberAsParticipant(Member memory member, CommitteeInfo memory _committeeInfo) private {
+		participants.length = Math.max(participants.length, _committeeInfo.freeParticipantSlotPos + 1);
```

Comment:
Then we may want to `assert(freeSlotPos <= length)`. wdyt?

Reply:
the assert is optional in both flavors. since we avoid asserting in other cases because "its internal and could only be caused by a bug" i don't see the reason to act differently here. lets have a general discussion about assertions?